### PR TITLE
Implement 2FA for Umbrel-Manager

### PR DESCRIPTION
**Pull request by dsbaars**
Implementing 2FA using TOTP as requested in  getumbrel/umbrel#985 and getumbrel/umbrel#1078.
Also needs modified version of umbrel-dashboard, related pull request getumbrel/umbrel-dashboard#395

### CODE REVIEW

File: logic/auth.js [hunk 1]
```diff
@@ -51,33 +68,58 @@ async function changePassword(currentPassword, newPassword, jwt) {
...
 
-      // update user file
-      await diskLogic.writeUserFile({ ...user, password: credentials.password, seed: encryptedSeed });
+        // update user file
+        await diskLogic.writeUserFile({ ...user, password: credentials.password, seed: encryptedSeed });
 
-      // update system password
-      await setSystemPassword(newPassword);
+        // update system password
+        await setSystemPassword(newPassword);
```
Comment (AaronDewes) [inline logic/auth.js:83]: Maybe reverting these changes would help getting this merged. I recommend to not lint any of your PRs, Umbrel hasn't used a configured linter in any project for over a year now, even though that could've prevented some bugs.

File: middlewares/auth.js [hunk 2]
```diff
@@ -136,7 +157,10 @@ async function accountJWTProtected(req, res, next) {
       if (error || user === false) {
         return next(new NodeError('Invalid JWT', 401)); // eslint-disable-line no-magic-numbers
       }
+
       req.logIn(user, function (err) {
+
+
```
Comment (AaronDewes) [inline middlewares/auth.js:163]: Please revert this too to keep changes minimal.

File: middlewares/auth.js [hunk 3]
```diff
@@ -106,6 +116,17 @@ function basic(req, res, next) {
...
 
+        // check 2FA token when enabled
+        if(userData.totpEnabled) {
+          let vres = notp.totp.verify(req.body.totpToken, userData.totpKey)
+    
+          if(vres && vres.delta == 0) {
+            
+          } else {
+            return next(new NodeError('Unable to authenticate', 401)); 
+          }
```
Comment (AaronDewes) [inline middlewares/auth.js:127]:
```suggestion
          if(!vres || vres.delta == 0) {
            return next(new NodeError('Unable to authenticate', 401)); 
          }
```

**Review by AaronDewes — COMMENTED** [review]
Just a few suggestions

**Comment by lukechilds** [comment]
Hey thanks for working on this @dsbaars!

We decided to go for a slightly different approach:

- https://github.com/getumbrel/umbrel-manager/pull/112
- https://github.com/getumbrel/umbrel-dashboard/pull/397

The passport-totp module doesn't really fit well with our existing codebase since we don't use sessions for authentication. Also we wanted to build 2FA into the existing authentication endpoints as opposed to adding new ones, along with doing a custom UI for it.

Really appreciate you taking the time to put this together though!

**Pull request closed without merging**